### ISSUE TITLE
Duplication in the Europarl dataset

### ISSUE THREAD

**Issue by chenyangh** [nlp-dke/NMTGMinor#4]
Hello, 

Thanks for the prompt response to my previous post.

I got some very high BLEU scores on the Europarl datasets, and it was too good to be true.

Therefore, I analyzed the duplication of the `europarl-full` variant. However, I found ~75% of the test data can be found in the training set. This holds for every EN-XX direction, and those are exact matches. 

I might make some mistakes tho, but here is the code I used to check `da-en`

```
x_lang = "da"
train_src_lines = open(f'europarl/ori_data/train/{x_lang}-en.s').readlines()
train_tgt_lines = open(f'europarl/ori_data/train/{x_lang}-en.t').readlines()


test_src_lines = open(f'europarl/ori_data/test/tok/{x_lang}-en.multiway.s').readlines()
test_tgt_lines = open(f'europarl/ori_data/test/tok/en-{x_lang}.multiway.s').readlines()

dup_count = 0
for src, tgt in zip(test_src_lines, test_tgt_lines):
    try:
        ind = train_src_lines.index(src)
    except ValueError:
        ind = -1

    if ind != -1:
        test_tgt_strip = tgt.strip()
        train_tgt_strip_langtok = ' '.join(train_tgt_lines[ind].split()[1:])
        if test_tgt_strip == train_tgt_strip_langtok:
            dup_count += 1

print(x_lang, dup_count, len(test_src_lines))

```

The output is  `da 1428 2000`

For every other en-xx pair, I got
```
da 1428 2000
de 1315 2000
es 1573 2000
fi 1212 2000
fr 1387 2000
it 1436 2000
nl 1490 2000
pt 1551 2000
```

**Comment by dannigt** [comment]
Thanks a lot for reporting the issue!

You are right. After checking I see the training set of `Europarl-full` isn't clean - due to my mistake when pointing the original data from MMCR4NLP.

The mistake impacted `Europarl-full`. I've verified that for `Europarl-non-overlap` and `Europarl-multiway` the training set is clean.

In the experiments on `Europarl-full`, we also included the zero-shot directions in the dev set for early stopping (reported in the 3rd paragraph of [section 3.2](https://arxiv.org/pdf/2012.15127.pdf)). As a side effect of this, the problem with the contaminated training set didn't pop up earlier.

I'm starting to retrain a model on the clean data for `Europarl-full`; I'll also upload the clean preprocessed data asap. Will keep you posted about the progress.

Thanks again for noticing this, and my apologies for the inconvenience.

**Comment by chenyangh** [comment]
Thanks for getting back to me so soon. I think this leaking might not affect the zero-shot experiments too much, that being said, we need to switch to the clean dataset. I am recently trying to replicate your approach using Fairseq, but I haven't actually replicated the numbers. Here are my training hyper-parameters that I find might be different from yours 
```
lr: 0.0005
max update: 100K
batch size: 32K tokens
```
In your training script, the LR is set to 2, is it literally 2?

BTW, I was trying to preprocess `Europarl-non-overlap` and `Europarl-multiway `, and I found that they are the same, and it looks like they are only different by using different BPE models. 
The zip file for `Europarl-full` is about 2.1G, but both `non-overlap` and `multiway` are 255M. I was expecting `non-overlap` to be the size of ~ 1.8G if I understand the setting correctly.

I am using Fairseq to replicate your results, so this might be a little tricky. Sorry to bother you a lot.

**Comment by dannigt** [comment]
Thanks a lot for following up!

**RE learning rate:** 

The learning rate of 2.0 isn't the exact value of the LR - it is a scaling factor on top of the LR schedule from the Transformer paper (section 5.3 of https://arxiv.org/pdf/1706.03762.pdf). In this implementation it's initialized here: https://github.com/nlp-dke/NMTGMinor/blob/f3ae5728e4b7ba4544ba06e3754d026f9a0cf9ae/onmt/optim.py#L424 and updated here during training: https://github.com/nlp-dke/NMTGMinor/blob/f3ae5728e4b7ba4544ba06e3754d026f9a0cf9ae/onmt/optim.py#L495-L498
It should behave similarly to the original [tensor2tensor implementation](https://github.com/tensorflow/tensor2tensor/blob/ef1fccebe8d2c0cf482f41f9d940e2938c816c78/tensor2tensor/utils/learning_rate.py#L105-L107), although they seem to use a much bigger scaling factor.
For Fairseq I got comparable results by `--lr-scheduler inverse_sqrt --warmup-updates 4000 --warmup-init-lr 1e-7` on the IWSLT dataset (detailed later).

**RE data size:**

Both `Europarl-non-overlap` and `Europarl-mutliway` are subsets of `Europarl-full` (~1/8 of the size). A detailed description of how the subsets are constructed is [here](https://github.com/nlp-dke/NMTGMinor/issues/1#issuecomment-919745030). With this experiment setup, we investigate the effect of multiway data on zero-shot performance, holding the data size equal.  (`Europarl-non-overlap` contains no multiway data at all while `Europarl-multiway` contains only multiway data.)
To be more precise I should rename them to  `Europarl-subset-non-overlap` and `Europarl-subset-mutliway`. :)

**RE fairseq:**

So far I've reproduced the results on the IWSLT dataset, with detailed training script [here](https://github.com/nlp-dke/NMTGMinor/issues/1#issuecomment-920623762) and [here](https://github.com/nlp-dke/NMTGMinor/issues/1#issuecomment-920627137). We did notice that it was important to use `--encoder-langtok "tgt"` instead of `--encoder-langtok "src"` for the residual-drop approach to work. One (untested) hypothesis is that prepending src tokens makes the source sentences more language-specific and therefore more difficult for zero-shot inference. 

Just want to let you know your feedback and questions are really appreciated! :)

**Comment by chenyangh** [comment]
Thanks a lot! 👍 

Yes, I am also using `--encoder-langtok "tgt"` and the decoder langtok is omitted. [Wu et al. (2021)](https://aclanthology.org/2021.findings-acl.264.pdf) also suggest this setting is better for the zero-shot setting.

**Comment by chenyangh** [comment]
Hi, I have just replicated the results on the IWSLT. 🎉 

I was wrong by not using the decoder `langtok`, which I should have realized earlier.

**Comment by dannigt** [comment]
Just an update that the corrected Europarl data is here: https://github.com/nlp-dke/NMTGMinor/tree/master/recipes/zero-shot#download-preprossed-data

We also uploaded the original data [here](https://github.com/nlp-dke/NMTGMinor/tree/master/recipes/zero-shot#download-original-data), to make it easier with different preprocessing pipelines.

I will follow up with results once training is finished on the corrected data.
